### PR TITLE
[Aichat] Handle user input too large error

### PR DESCRIPTION
**Pull request by fisher-alice**
This PR adds handling for the "input validation error: `inputs` must have less than 3000 tokens."
A new code (`USER_INPUT_TOO_LARGE: 1005`) was added to `AI_REQUEST_EXECUTION_STATUS` for this error. This status is set to `1005` in `aichat_request_chat_completion_job`  if the exception's message includes the error description.

Screenshot 2024-09-05 at 3 57 21 PM

On the frontend, the error message has been updated to help the user understand the error and how to resume the conversation with the chatbot:
Screenshot 2024-09-05 at 3 41 13 PM

## Links
[jira](https://codedotorg.atlassian.net/browse/LABS-833)
[Slack thread](https://codedotorg.slack.com/archives/C06FELVTV0Q/p1715981040002769)

## Testing story
Added a dashboard unit test and tested locally on an /allthethings level.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/jobs/aichat_request_chat_completion_job.rb [hunk 1]
```diff
@@ -18,7 +18,12 @@ class AichatRequestChatCompletionJob < ApplicationJob
     end
 
     request = arguments.first[:request]
-    request.update!(response: exception.message, execution_status: SharedConstants::AI_REQUEST_EXECUTION_STATUS[:FAILURE])
+    execution_status_code = SharedConstants::AI_REQUEST_EXECUTION_STATUS[:FAILURE]
+    if exception.message.include? "must have less than 3000 tokens"
```
Comment (fisher-alice) [inline dashboard/app/jobs/aichat_request_chat_completion_job.rb:22]: Spent some time looking for a way to parse `exception` other searching for substring in `message` - printed out `exception.methods` and didn't see anything that would point to this type of error. Open to ideas!

File: apps/i18n/common/en_us.json [hunk 2]
```diff
@@ -162,6 +162,7 @@
   "aiChatResponseError": "There was an error getting a response. Please try again.",
   "aiChatRateLimitError": "You've sent too many messages in the last minute. Please wait a moment before sending another message.",
   "aiChatTooPersonalUserMessage": "This message has been flagged as too personal.",
+  "aiChatUserInputTooLargeMessage": "AI Chat can only process so much information and your last message used too many words. Please clear the chat or shorten your last message to continue chatting.",
```
Comment (fisher-alice) [inline apps/i18n/common/en_us.json:165]: Capitalized the 'C' in `aiChat` for consistency but it hurt me to do so.

File: apps/src/aiComponentLibrary/chatMessage/ChatMessage.tsx [hunk 3]
```diff
@@ -28,6 +28,7 @@ const ChatMessage: React.FunctionComponent<ChatMessageProps> = ({
 
   const hasDangerStyle =
     status === Status.PROFANITY_VIOLATION ||
+    status === Status.USER_INPUT_TOO_LARGE ||
```
Comment (sanchitmalhotra126) [inline apps/src/aiComponentLibrary/chatMessage/ChatMessage.tsx:31]: Thought for later - might be good to move this logic out of the shared `ChatMessage` component and into aichat logic, and just pass in a string with the desired error message here. That way the different chat applications don't necessarily need to have the same set of logical statuses and this component can be more presentation focused.

File: dashboard/app/jobs/aichat_request_chat_completion_job.rb (hunk 1 above)
Comment (sanchitmalhotra126) [reply dashboard/app/jobs/aichat_request_chat_completion_job.rb:22]: Is the type of returned exception any more specific than the standard `Exception`? If so, I think we could add a separate rescue_from (i.e. `rescue_from AwsException do |exception|` if something like that exists) and handle this there.

**Review by sanchitmalhotra126 — APPROVED** [review]
LGTM! One suggestion for exception handling if it's possible

File: apps/src/aiComponentLibrary/chatMessage/ChatMessage.tsx (hunk 3 above)
Comment (fisher-alice) [reply apps/src/aiComponentLibrary/chatMessage/ChatMessage.tsx:31]: Sounds good - i'll create a jira ticket for this.

Comment (fisher-alice) [reply apps/src/aiComponentLibrary/chatMessage/ChatMessage.tsx:31]: https://codedotorg.atlassian.net/browse/LABS-997

File: dashboard/app/jobs/aichat_request_chat_completion_job.rb (hunk 1 above)
Comment (fisher-alice) [reply dashboard/app/jobs/aichat_request_chat_completion_job.rb:22]: Hey Sanchit - documenting what I attempted even though it unfortunately did not work.

In the server output as it's currently implemented the following is printed:
```
AichatRequestChatCompletionJob Error: /home/ubuntu/.rbenv/versions/3.0.5/lib/ruby/gems/3.0.0/gems/aws-sdk-core-3.201.3/lib/seahorse/client/plugins/raise_response_errors.rb:17:in `call': Received client error (422) from primary with message "{"error":"Input validation error: `inputs` tokens + `max_new_tokens` must be <= 4096. Given: 3758 `inputs` tokens and 512 `max_new_tokens`","error_type":"validation"}". See https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logEventViewer:group=/aws/sagemaker/Endpoints/gen-ai-mistral-7b-inst-v01 in account 475661607190 for more information. (Aws::SageMakerRuntime::Errors::ModelError)
	from /home
```
Thus, I added the following, similar to the [aws error rescue](https://github.com/code-dot-org/code-dot-org/blob/dff352317734de122d1246763ec5efa7408eb3cc/dashboard/app/jobs/concerns/ai_rubric_config.rb#L123) in `ai_rubric_config`:
```
  rescue Aws::SageMakerRuntime::Errors::ModelError => exception
    if rack_env?(:development)
      puts "AichatRequestChatCompletionJob Error: #{exception.full_message}"
    end

    request = arguments.first[:request]
    execution_status_code = SharedConstants::AI_REQUEST_EXECUTION_STATUS[:FAILURE]
    if exception.message.include? "must have less than 3000 tokens"
      execution_status_code = SharedConstants::AI_REQUEST_EXECUTION_STATUS[:USER_INPUT_TOO_LARGE]
    end

    request.update!(response: exception.message, execution_status: execution_status_code)
    Honeybadger.notify(
      "AichatRequestChatCompletionJob failed with unexpected error: #{exception.message}",
      context: {
        request: request.to_json
      }
    )

    # Raise an exception to notify our system of the failed job.
    raise "AichatRequestChatCompletionJob failed with unexpected error: #{exception.message}. Context: #{request.to_json}"
```
However, I got the following output and `1005` was NOT returned.
` [2024-09-06 18:43:59.152] ERROR -- NotImplementedError: nil`

I tried requiring `aws-sdk-sagemakerruntime` in the file, but then got the following error:
```
NameError at /aichat/start_chat_completion
==========================================
undefined local variable or method `‘aws’ for main:Object
```
I also tried using `rescue_from Aws::SageMakerRuntime::Errors::ModelError do |exception|` and kept the other rescue in. I got status `1000` back. The `rescue` was from the general exception rescue, not the more specific `Aws` error.

I will keep as-is for now, but am definitely open to a follow-up if you have any advice. Thanks!

**Pull request merged**